[![codecov](https://codecov.io/gh/bioinfologics/bsg/branch/master/graph/badge.svg)](https://codecov.io/gh/bioinfologics/bsg)
[![Build Status](https://travis-ci.org/bioinfologics/bsg.svg?branch=master)](https://travis-ci.org/bioinfologics/bsg)

# bsg
The **B**asic **S**equence **G**raph framework.

BSG allows you to work with sequence graphs such as those generated by various genome assemblers.

It provides a workspace that can contain a graph and datastores for paired, linked and long reads. These reads can be mapped to the graph.

A SWIG API enables BSG to be used as a Python module, and there is experimental Julia and R support.

The documentation is severily lacking right now. We're working on it.

There is [BSG code documentation](https://bioinfologics.github.io/bsg/), although again it is lacking right now.

# Using bsg (graph simplification/scaffolding using datastores and existing heuristics)

**Warning: BSG is undergoing major changes and this master repository is outdated. I will be updated soon with a working version.**

The current version of needs as input:
- A GFA+fasta graph of the current (input) assembly. Preferably a w2rap-contigger contigs_raw.gfa graph assembled at k=200.
- An illumina dataset to use as k-mer spectrum for the KCI. This is usually the dataset that generated the input assembly graph.
- 10x Chromium Illumina and/or LMP and/or Pacbio and/or Nanopore reads.

All read sequence files need to be converted to datastores. These are
indexed files BSG can use to access every read's sequence
efficiently.

A full workspace must be created, with all of these components, this can later save other
pre-computed status data such as read mapping positions.

Some example commands (**warning, BSG is being thoroughly updated right now, this may only be useful as a trivial exaple now**):

```
#create new workspace with graph
bsg-workspace make -g {input.gfa} -o {wsname}

#create new datastore with linked reads
bsg-datastore make -t 10x -1 {10x_R1.fastq} -2 {10x_R2.fastq} -o {dsname.bsgds}

#add 10x datastore to ws
bsg-workspace add -w {wsname.bsgws} -l {dsname.bsgds}

#add KCI spectrum computed from PE files into workspace
bsg-kmerspectrum make -w {wsname.bsgws} -f {r1.fastq} -f {r2.fastq}

#map reads to graph and save ws to new _mapped file
bsg-mapper -w {wsname.bsgws} -o {wsname_mapped}
```
